### PR TITLE
bfsos - local file system adapter

### PR DESCRIPTION
**Pull request by mxmCherry**

### CODE REVIEW

File: bfsos/atomic_file.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+// Close commits the file.
+func (f *atomicFile) Close() error {
+	defer f.cleanup()
+
+	if err := f.File.Close(); err != nil {
+		return err
+	}
+
+	if err := os.MkdirAll(filepath.Dir(f.name), 0755); err != nil {
```
Comment (dim) [inline bfsos/atomic_file.go]: needs to be 0777, umask will kick in anyway

File: bfsos/atomic_file.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package bfsos
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+// atomicFile represents a file, that's written only on Close.
+type atomicFile struct {
```
Comment (dim) [inline bfsos/atomic_file.go]: why did you decide to roll your own in the end?

File: bfsos/bfsos_test.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+var _ = Describe("Bucket", func() {
+	var rootDir, tmpDir string
+	var data = lint.Data{}
+
+	BeforeEach(func() {
+		if os.Getenv("BFSOS_TEST") == "" {
+			Skip("test is disabled, enable via BFSOS_TEST")
+		}
+
+		Expect(os.MkdirAll(tmpRoot, 0755)).To(Succeed())
```
Comment (dim) [inline bfsos/bfsos_test.go]: can we do this instead:

```
tmpRoot, err = ioutil.TempDir("", "bfsfs")
...
rootDir = filepath.Join(tmpRoot, "root")
tmpDir = filepath.Join(tmpRoot, "tmp")
```

File: bfsos/bfsos_test.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	"testing"
+
+	"github.com/bsm/bfs/bfsos"
+	"github.com/bsm/bfs/testdata/lint"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
```
Comment (dim) [inline bfsos/bfsos_test.go]: pls remove the two, I don't like tests to write files to local dirs, always create a brand new base dir via `ioutil.TempFile("", "bfsfs")`

File: bfsos/bfsos_test.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "bfs/bfsos")
+}
+
+var _ = BeforeSuite(func() {
+	if os.Getenv("BFSOS_TEST") == "" {
+		return
+	}
+
+	Expect(os.MkdirAll(filepath.Join(readonlyDir, "readonly"), 0755))
```
Comment (dim) [inline bfsos/bfsos_test.go]: same thing:

```
readonlyDir, err = ioutil.TempDir("", "bfsfs-readonly")
```

File: bfsos/bfsos_test.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		Expect(ioutil.WriteFile(name, []byte(name), 0644)).To(Succeed())
+	}
+})
+
+var _ = AfterSuite(func() {
+	if os.Getenv("BFSOS_TEST") == "" {
+		return
+	}
+
+	Expect(os.RemoveAll(tmpRoot))
```
Comment (dim) [inline bfsos/bfsos_test.go]: `Expect(os.RemoveAll(readonlyDir)).To(Succeed())`

File: bfsos/bucket.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	}, nil
+}
+
+// Glob lists the files mathing a glob pattern.
+func (b *bucket) Glob(_ context.Context, pattern string) (bfs.Iterator, error) {
+	if pattern == "" { // would return just current dir
+		return newIterator(nil), nil
+	}
+
+	matches := make([]string, 0, 10)
```
Comment (dim) [inline bfsos/bucket.go]: do `var matches []string`, don't allocate

File: bfsos/bucket.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+	// filepath.Glob is not suitable, returns dirs as well:
+	err := filepath.Walk(b.root, func(path string, info os.FileInfo, err error) error {
```
Comment (dim) [inline bfsos/bucket.go]: please use Glob and then os.Stat on each to see if it's a dir

File: bfsos/bucket.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+	if err != nil {
+		return nil, normError(err)
+	}
+	return f, nil
+}
+
+// Remove removes a object.
+func (b *bucket) Remove(ctx context.Context, name string) error {
+	err := os.Remove(b.resolve(name))
+	if err != nil && os.IsExist(err) {
```
Comment (dim) [inline bfsos/bucket.go]: i don't think this is the right logic, you should return nil on NotExist and err on all other errors

File: bfsos/bucket.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+	return nil
+}
+
+// Close closes the bucket.
+func (b *bucket) Close() error {
+	return nil // noop
+}
+
+// Copy supports copying of objects within the bucket.
+func (b *bucket) Copy(ctx context.Context, srcName, dstName string) error {
```
Comment (dim) [inline bfsos/bucket.go]: please remove `Copy`, the generic copy method does pretty much exactly that, you are not saving something

File: bfsos/bucket.go [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+	if _, err := io.Copy(dst, src); err != nil {
+		cancel()
+		return normError(err)
+	}
+	return normError(dst.Close())
+}
+
+// resolve returns full safely rooted path.
+func (b *bucket) resolve(name string) string {
+	return filepath.Join(b.root, filepath.Join("/", name))
```
Comment (dim) [inline bfsos/bucket.go]: shouldn't `filepath.Join(b.root, name)` do?

File: bfsos/iterator.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
+package bfsos
+
+import (
+	"sync"
+)
+
+// iterator implements an iterator over file path list.
+type iterator struct {
+	mutex sync.Mutex
```
Comment (dim) [inline bfsos/iterator.go]: remove mutex. iterators are not supposed to be thread-safe, use same method as inmem one

File: bfsos/atomic_file.go (hunk 2 above)
Comment (mxmCherry) [reply bfsos/atomic_file.go]: https://github.com/dchest/safefile doesn't support custom tmp dir.

So our tests fail, as it creates file in current dir.

Theoretically, I can PR it with `safefile.Create("file.bin", 0644, safefile.WithTempDir("testdata/tmp"))`, but decided it'll take longer than doing a quick internal atomicFile.

File: bfsos/bucket.go (hunk 9 above)
Comment (mxmCherry) [reply bfsos/bucket.go]: Well, that does what you describe.

Non-nil error, and if it's IsExist - then return err.

Non-nil error, and it's IsNotExist - return nil.

Comment (mxmCherry) [reply bfsos/bucket.go]: Let me change this for readability.

File: bfsos/bucket.go (hunk 11 above)
Comment (mxmCherry) [reply bfsos/bucket.go]: Probably, I'm just being too defensive, but my variant protects (roots properly) from `b.Method("../../../../../../etc/passwd")`, better leave it, it's safer (mistypes etc).

File: bfsos/atomic_file.go (hunk 2 above)
Comment (mxmCherry) [reply bfsos/atomic_file.go]: `safefile` does just os.Rename, as you've already checked.

Leaving custom `atomicFile` implementation, at least for now, doesn't hurt anyone.

File: bfsos/bucket.go (hunk 10 above)
Comment (mxmCherry) [reply bfsos/bucket.go]: :man_facepalming: sorry, haven't checked

**Comment by mxmCherry** [comment]
Let me ask again, so it won't be lost in review comments - probs, better rename `bfsOS` -> `bfsFS`?..

File: bfsos/bucket.go [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+	if err != nil {
+		return nil, normError(err)
+	}
+	return f, nil
+}
+
+// Remove removes a object.
+func (b *bucket) Remove(ctx context.Context, name string) error {
+	err := os.Remove(b.resolve(name))
+	if err != nil {
```
Comment (dim) [inline bfsos/bucket.go]:
```
if err != nil && !os.IsNotExist(err) {
  return err
}
return nil
```
minor

File: bfsos/bucket_test.go [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+		data lint.Data
+	)
+
+	BeforeEach(func() {
+		if os.Getenv("BFSOS_TEST") == "" {
+			Skip("test is disabled, enable via BFSOS_TEST")
+		}
+
+		var err error
+		tmp, err = ioutil.TempDir("", "bfsos")
```
Comment (dim) [inline bfsos/bucket_test.go]: here's what I meant:

```
var tempDir string

var _ = BeforeSuite(func() {
  if os.Getenv("BFSOS_TEST") == "" { ... }

  var err error
  tempDir, err = ioutil.TempDir("", "bfsos")
  Expect(err).NotTo(HaveOccurred())

  readonlyDir := filepath.Join(tempDir, "readonly")
  Expect(os.MkdirAll(  readonlyDir, 0777)).To(Succeed())
  populateReadonlyFiles(readonlyDir) // <- you can inline that one here
})
var _ = AfterSuite(func() {
  if tempDir != "" {
    Expect(os.RemoveAll(tempDir)).To(Succeed())
  }
})

var _ = Describe("Bucket", func() {
  var subjectDir string

  BeforeEach(func() {
    if os.Getenv("BFSOS_TEST") == "" { ... }
    subjectDir = filepath.Join(tempDir, "subject")
    Expect(os.MkdirAll(subjectDir).To(Succeed())
  })
  AfterEach(func() { ... })
  ...
```

File: bfsos/bucket_test.go [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+		subjectDir := filepath.Join(tmp, "subject")
+		Expect(os.MkdirAll(subjectDir, 0777)).To(Succeed())
+
+		readonlyDir := filepath.Join(tmp, "readonly")
+		Expect(os.MkdirAll(readonlyDir, 0777)).To(Succeed())
+
+		tmpDir := filepath.Join(tmp, "tmp")
+		Expect(os.MkdirAll(tmpDir, 0777)).To(Succeed())
+
+		populateReadonlyFiles(readonlyDir)
```
Comment (dim) [inline bfsos/bucket_test.go]: then you don't need to write 2k files on every BeforeEach

File: bfsos/bucket_test.go [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+)
+
+var _ = Describe("Bucket", func() {
+	var (
+		tmp  string
+		data lint.Data
+	)
+
+	BeforeEach(func() {
+		if os.Getenv("BFSOS_TEST") == "" {
```
Comment (dim) [inline bfsos/bucket_test.go]: actaully, I don't think we need this. this test can run on any env, it doesn't need an env guard

**Pull request merged**